### PR TITLE
Add some options for textutils.serialize

### PR DESCRIPTION
**Pull request by magiczocker10**
This pull request adds some small options to change the behavior and output of textutils.serialize.

I have first issued this in #636 and was asked to create a pull request with some requested changes to it.

### CODE REVIEW

File: src/main/resources/data/computercraft/lua/rom/apis/textutils.lua [hunk 1]
```diff
@@ -262,9 +262,16 @@ local g_tLuaKeywords = {
     ["while"] = true,
 }
 
-local function serializeImpl(t, tTracking, sIndent)
-    local sType = type(t)
+local function serializeImpl(t, tTracking2, sIndent, opts)
+    local sType, tTracking = type(t), {}
     if sType == "table" then
+				if opts.advancedComparing then
```
Comment (SquidDev) [inline src/main/resources/data/computercraft/lua/rom/apis/textutils.lua:268]: Instead of copying the table, could we just do `tTracking[t] = nil` at the end of this block (and probably move `tTracking[t] = nil` inside the `else` branch).

File: src/main/resources/data/computercraft/lua/rom/apis/textutils.lua [hunk 2]
```diff
@@ -275,20 +282,20 @@ local function serializeImpl(t, tTracking, sIndent)
             return "{}"
         else
             -- Other tables take more work
-            local sResult = "{\n"
-            local sSubIndent = sIndent .. "  "
+            local sResult = "{" .. (opts.noWrap and "" or "\n")
```
Comment (SquidDev) [inline src/main/resources/data/computercraft/lua/rom/apis/textutils.lua:285]: I wonder if it's worth hoisting the "separator string" logic into one place. So something like:

```lua
local open, close, equal, comma = "{\n", "\n}", " = ", ",\n"
-- I'm sure there's more ("[ ", " ] = ", etc...), but hopefully gives the right idea
if ops.noWhiteSpace then
  open, close, equal, comma = "{", "}", "=", ","
end
```

I'd personally also merge the `noWhiteSpace` and `noWrap` options. I guess I'm seeing two use cases for `textutils.serialize`:
 - Want the output to be easy to read (e.g. printing, configuration files).
 - Want the output to be compact (e.g. saving large amounts of data to a file).

File: src/main/resources/data/computercraft/lua/rom/apis/textutils.lua [hunk 3]
```diff
@@ -645,9 +652,17 @@ end
 -- @throws If the object contains a value which cannot be
 -- serialised. This includes functions and tables which appear multiple
 -- times.
-function serialize(t)
+function serialize(t, opts)
     local tTracking = {}
-    return serializeImpl(t, tTracking, "")
+		expect(2, opts, "table", "nil")
+		if opts then
+			field(opts, "noWrap", "boolean", "nil")
```
Comment (SquidDev) [inline src/main/resources/data/computercraft/lua/rom/apis/textutils.lua:659]: Possibly worth switching to snake_case - that's what we do with `unserializeJSON`.

I'd also make all the terms positive, so you'd have `{ wrap = false }` instead of `{ noWrap = true }`. It gets a little confusing if you're specifying options explicitly - the double negative of `{ noWrap = false }` is a little hard to parse :).

**Review by SquidDev — CHANGES_REQUESTED** [review]
Thank you for submitting this, really sorry for the slow review! Just a couple of comments:

File: src/main/resources/data/computercraft/lua/rom/apis/textutils.lua (hunk 1 above)
Comment (magiczocker10) [reply src/main/resources/data/computercraft/lua/rom/apis/textutils.lua]: Done

**Review by SquidDev — APPROVED** [review]

**Comment by SquidDev** [comment]
Sorry for taking so long to get to this. Thank you very much! 

Doc examples on the website won't work quite yet as copy-cat hasn't updated yet.
